singularity-eos/eos/eos_gruneisen.cpp

```diff
@@ -12,12 +12,155 @@
+  Real discriminant, root1, root2;
+  if (s2 == 0 && s3 == 0 && s1 > 0) {
+    // Linear Us-up analytic root
+    root = 1 / s1;
+  } else if (s3 == 0 && s2 != 0) {
+    // Quadratic Us-up
+    discriminant = square(s1) + 4 * s2;
+    if (discriminant < 0) {
+      root = -1; // Imaginary roots, so no limit
+    } else {
+      root1 = (s1 + std::sqrt(square(s1) + 4 * s2)) / (-2 * s2);
+      root2 = (s1 - std::sqrt(square(s1) + 4 * s2)) / (-2 * s2);
+      root = find_min_bounded_val(root1, root2, 0, 1);
+    }
+  } else if (s3 != 0) {
+    // Cubic Us-up
+    discriminant = -18 * s3 * s2 * s1 - 4 * cube(-s2) + square(s2) * square(s1) -
+                   4 * s3 * cube(s1) - 27 * square(s3);
+    // Use discriminant to help find roots
+    if (discriminant == 0) {
+      // Easy analytical roots (probably not the case)
+      if (square(s2) == 3 * s3 * s1) {
+        // Single root with multiplicity of 3 (inflection point is root)
+        // Also guards against divide by zero
+        root = -s2 / 3 / s3;
+      } else {
+        // One root has multiplicity of 2 (local extremum is a root)
+        root1 = (-9 * s3 - s2 * s1) / 2 / (square(s2) - 3 * s3 * s1);
+        root2 = (-4 * s3 * s2 * s1 - 9 * square(s3) + cube(s2)) /
+                (-s3 * (square(s2) - 3 * s3 * s1));
+        root = find_min_bounded_val(root1, root2, 0, 1);
+      }
+    } else {
+      // We'll need an iterative method to search for the minimum bounded root
+      Real minbound = 0;
+      Real maxbound = 1;
+      if (discriminant > 0) {
+        // Three real roots. We need to use the extrema to ensure we have a proper bracket
+        // in which to search for the root.
+        const Real extremum1 =
+            (2 * s2 + std::sqrt(square(-2 * s2) - 4 * 3 * s3 * s1)) / (-2 * 3 * s3);
+        const Real extremum2 =
+            (2 * s2 - std::sqrt(square(-2 * s2) - 4 * 3 * s3 * s1)) / (-2 * 3 * s3);
+        const Real min_extremum = std::min(extremum1, extremum2);
+        const Real max_extremum = std::max(extremum1, extremum2);
+        if (s3 < 0) {
+          // Cubic is *increasing*
+          // Because poly(eta = 0) = 1, the only possible root for an increasing function
+          // will lie between the etrema.
+          minbound = std::max(min_extremum, minbound);
+          maxbound = std::min(std::max(max_extremum, minbound), maxbound);
+        } else {
+          // Cubic is *decreasing*
+          // Because poly(eta = 0) = 1, the only possible root for a decreasing function
+          // will lie outside of the extrema. Further, the only possibility for multiple
+          // bound roots will occur when the extremum are both positive
+          if (min_extremum > 0) {
+            maxbound = std::min(min_extremum, maxbound);
+          }
+        }
+      }
+      if (poly(minbound) * poly(maxbound) <= 0) {
+        // Root is appropriately bounded
+        using RootFinding1D::bisect;
```

Comment:
I might try the new regula falsi method instead of bisection, thought it probably doesn't matter too much

Reply:
Yeah it wasn't available yet when I was coding this up. I just wanted to avoid secant really. I'll change this to the false position root finder